<a href="https://colab.research.google.com/github/ayushhbroshan/Projects/blob/main/IMDB_Movies_analysis_by_ayush.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [31]:
# @title
from IPython.display import display, HTML

display(HTML("""
<div style="
    background: linear-gradient(135deg, #111111 0%, #242424 55%, #f5c518 100%);
    padding: 45px 35px;
    border-radius: 18px;
    color: white;
    text-align: center;
    box-shadow: 0 8px 25px rgba(0,0,0,0.35);
    margin-bottom: 25px;
">

    <div style="
        display: inline-block;
        background-color: #f5c518;
        color: #111111;
        font-size: 30px;
        font-weight: bold;
        padding: 8px 18px;
        border-radius: 6px;
        margin-bottom: 18px;
    ">
        IMDb
    </div>

    <h1 style="
        font-size: 42px;
        margin: 10px 0;
        color: white;
    ">
        🎬 IMDb Movie Analysis
    </h1>

    <p style="
        font-size: 20px;
        margin-top: 15px;
        color: #eeeeee;
    ">
        Exploring Movies, Ratings, Genres & Cinema Trends
    </p>

    <div style="
        margin: 30px auto 10px auto;
        max-width: 850px;
        background: rgba(0,0,0,0.55);
        padding: 25px;
        border-radius: 12px;
    ">

        <h2 style="color:#f5c518;">
            🍿 Project Overview
        </h2>

        <p style="
            font-size: 17px;
            line-height: 1.7;
            color: #f1f1f1;
        ">
            This project explores the IMDb Movies dataset to understand the
            characteristics and trends that shape the world of cinema.
            Through data cleaning, exploratory data analysis and visualization,
            we investigate movie ratings, genres, release years, runtime,
            popularity and other important attributes.
        </p>

        <p style="
            font-size: 17px;
            line-height: 1.7;
            color: #f1f1f1;
        ">
            The analysis aims to transform raw movie data into meaningful
            insights about audience perception and the changing landscape of
            films. By examining patterns in IMDb ratings and movie
            characteristics, we can better understand how different aspects
            of cinema are represented in the dataset.
        </p>

    </div>

    <p style="
        font-size: 16px;
        color: #dddddd;
        margin-top: 25px;
    ">
        🎥 Data Analysis &nbsp; | &nbsp; 📊 Visualization &nbsp; | &nbsp;
        ⭐ IMDb Ratings &nbsp; | &nbsp; 🍿 Cinema Trends
    </p>

</div>
"""))

In [11]:
# @title
import kagglehub
import os
import pandas as pd

path = kagglehub.dataset_download("ashishjangra27/imdb-movies-dataset")

print("Downloaded to:", path)
print("Files:", os.listdir(path))


Using Colab cache for faster access to the 'imdb-movies-dataset' dataset.
Downloaded to: /kaggle/input/imdb-movies-dataset
Files: ['movies.csv', 'Genre']


In [12]:
# @title
import os
import json
import re
import numpy as np
import pandas as pd
from tqdm.auto import tqdm

df = pd.read_csv(os.path.join(path, "movies.csv"))

df = df.dropna()

df.head()

/tmp/ipykernel_576/1157928283.py:8: DtypeWarning: Columns (7,8) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(os.path.join(path, "movies.csv"))


,id,name,year,rating,certificate,duration,genre,votes,gross_income,directors_id,directors_name,stars_id,stars_name,description
0,tt4710316,Best in Sex: 2015 AVN Awards,(2015 TV Special),4.0,TV-MA,94 min,"Adult, News",124.0,0,nm1624094,Gary Miller,"nm4766272,nm2670531,nm4920605,nm6284246","Farrah Laurel Abraham,Asa Akira,Anikka Albrite...",The hottest adult stars and top adult movies a...
1,tt1281857,Naughty Novelist,(2008 Video),3.8,Not Certified,88 min,Adult,174.0,0,nm0045256,John Bacchus,"nm0128986,nm1969196,nm0451160,nm6130462","Darian Caine,Jackie Stevens,A.J. Khan,Arrora",Darian is a successful journalist but when she...
2,tt2294954,2011 AVN Awards Show,(2011 TV Special),5.7,Not Certified,83 min,"Adult, News",39.0,0,"nm1624094,nm0754845","Gary Miller,Timothy E. Sabo","nm2200343,nm2670531,nm1267549,nm3585599","Aubrey Addams,Asa Akira,Monique Alexander,Rave...",Add a Plot
3,tt6843596,Best in Sex: 2017 AVN Awards,(2017 TV Special),4.9,TV-MA,87 min,"Adult, News",225.0,0,nm1624094,Gary Miller,"nm5221471,nm2670531,nm4920605,nm3038816","Amirah Adara,Asa Akira,Anikka Albrite,Britney ...","Considered the industry's highest honor, the A..."
4,tt3705604,AVN Awards 2014,(2014 TV Special),6.7,R,82 min,"Adult, News",101.0,0,nm1624094,Gary Miller,"nm2670531,nm4920605,nm6284246,nm3992720","Asa Akira,Anikka Albrite,August Ames,Jessie An...",Erotic filmmaking superstars Chanel Preston an...


### 2. Data Cleaning

#### 2.1) Cleaning name,duration, votes column

In [13]:
# @title
#name
df['name'] = df['name'].str.strip()
#duration
df['duration']=df['duration'].str.replace("min","").str.replace(",","")
#votes
df['votes'] = (
    df['votes']
    .astype(str)
    .str.replace(',', '', regex=False)
    .str.replace('.0', '', regex=False)
)

#### 2.2) Cleaning year Column

In [14]:
# @title
start_year = []
end_year   = []

for text in tqdm(df['year']):

    result = re.findall(r"\d{4}", text)

    if len(result) == 0:
        start_year.append('0')
        end_year.append('0')

    elif len(result) == 1:
        start_year.append(result[0])
        end_year.append(result[0])

    elif len(result) == 2:
        start_year.append(result[0])
        end_year.append(result[1])

df['start_year'] = start_year
df['start_year'] = df['start_year'].astype(int)

df['end_year']   = end_year
df['end_year'] = df['end_year'].astype(int)

  0%|          | 0/2590928 [00:00<?, ?it/s]

#### 2.3) clean gross income

In [15]:
# @title
df['gross_income'].value_counts()
df['gross_income']=df['gross_income'].str.replace(",","").str.replace("$","")
df['gross_income'] = (
    df['gross_income']
    .fillna('0')
    .str.upper()
    .str.replace('M', 'e6', regex=False)
    .str.replace('K', 'e3', regex=False)
    .astype(float)
)
df['gross_income']

,gross_income
0,0.0
1,0.0
2,0.0
3,0.0
4,0.0
...,...
2590927,0.0
2590928,0.0
2590929,0.0
2590930,0.0


#### 2.4) clean directors_id ,directors_name collumn

In [16]:
# @title
df['directors_id'] = (
    df['directors_id']
    .str.findall(r'nm\d+')
    .str.join(',')
)
df['directors_name'] = df['directors_name'].str.replace('nm0000000', 'Anonymous')
d_id_list=[]
d_name_list=[]
count=0
for d_id , d_name in df[['directors_id','directors_name']].astype(str).values:
    if len(d_id.split(','))!=len(d_name.split(',')):
        d_id_list.append(np.nan)
        d_name_list.append(np.nan)
    else:
        d_id_list.append(d_id)
        d_name_list.append(d_name)

df['d_id']=d_id_list
df['d_name']=d_name_list

df=df.dropna()
directors_dict=dict(zip(d_id_list,d_name_list))

with open('directors.json','w') as f:
    json.dump(directors_dict,f)

del d_id_list
del d_name_list
del df['directors_name']

#### 2.5) Cleaning stars_id and stars_name

In [17]:
# @title
df['stars_id']=df['stars_id'].str.replace('Anonymous','nm0000000')
df['stars_name']=df['stars_name'].str.replace('nm0000000','Anonymous')


star_id_list=[]
star_name_list=[]

for stars_id , stars_name in df[['stars_id','stars_name']].astype(str).values:
    if len(stars_id.split(','))!=len(stars_name.split(',')):
        star_id_list.append(np.nan)
        star_name_list.append(np.nan)
    else:
        star_id_list.append(stars_id)
        star_name_list.append(stars_name)

df['stars_id']=star_id_list
df['stars_name']=star_name_list

df=df.dropna()

stars_dict=dict(zip(star_id_list,star_name_list))

with open('stars.json','w') as f:
    json.dump(stars_dict,f)

del star_id_list
del star_name_list
del df['stars_name']

In [18]:
#### 2.6) deleting unecessory collumn

In [19]:
# @title
del df['year']
del df['description']
del df['d_id']
del df['d_name']

In [ ]:
df.head(5)

# 3.Time Trend Analysis of Movie Performance

## Business Context
The movie industry continuously evolves due to changes in audience preferences, content formats, technological advancements, and distribution platforms. Studying historical movie data allows analysts and industry stakeholders to understand how movie characteristics such as ratings, duration, genres, and audience engagement have changed over time.

Time trend analysis helps identify patterns in movie production, audience reception, and popularity across years, enabling better understanding of long-term developments within the entertainment industry.

---

## Problem Statement
Analyze how key movie attributes such as **ratings, votes, duration, and genre popularity** have changed over time in order to identify long-term trends in movie performance and audience engagement.

---

## Objectives
- Analyze trends in **movie production over time**
- Identify **changes in audience engagement using vote counts**
- Examine how **movie ratings evolve across years**
- Identify shifts in **genre popularity over time**
- Analyze whether **movie duration patterns change over time**

---

## Key Analytical Questions

### Production Trends
1. How has the **number of movies/series released each year** changed over time?  
2. Which years experienced the **highest movie/series production volume**?  
3. Are there periods showing **growth or decline in movie/series releases**?

### Audience Engagement Trends
4. How has the **average number of votes per movie/series changed over the years**?  
5. Which years show the **highest audience engagement** based on votes?

### Rating Trends
6. How have **average movie/series ratings changed over time**?  
7. Are newer movies/series receiving **higher or lower ratings compared to older movies**?

### Genre Trends
8. Which **genres were most popular in different years**?  
9. Are there noticeable **shifts in genre popularity over time**?

### Movie Characteristics
10. Has the **average movie duration changed across years**?

---

## Key Metrics
- **Number of movies released per year**
- **Average rating per year**
- **Total votes per year**
- **Average movie duration per year**
- **Genre frequency across years**

---

## Expected Insights
- Identification of **growth or decline periods in movie production**
- Changes in **audience engagement over time**
- Shifts in **genre popularity across different periods**
- Trends in **movie ratings across years**
- Patterns in **movie duration evolution**

---

## Deliverables
- **Time-series visualizations** showing yearly trends
- **Genre distribution charts across years**
- **Audience engagement trend analysis**
- **Summary of key findings and insights**
- **Interactive dashboard or analytical report**

#### 3.1 How many movies/series were released each year?
#### 3.2 Which years experienced the highest movie/series production volume?


In [20]:
# @title
import pandas as pd
import plotly.express as px

# -----------------------------
# Data Preparation
# -----------------------------

filtered_df = df[(df['start_year'] > 1900) & (df['start_year'] < 2024)]

movies_per_year = (
    filtered_df['start_year']
    .value_counts()
    .sort_index()
    .reset_index()
)

movies_per_year.columns = ['Year', 'Number of Movies']

# -----------------------------
# Calculate Numeric Insights
# -----------------------------

total_movies = filtered_df.shape[0]

start_year = filtered_df['start_year'].min()
end_year = filtered_df['start_year'].max()

years_span = end_year - start_year

peak_row = movies_per_year.loc[movies_per_year['Number of Movies'].idxmax()]
peak_year = int(peak_row['Year'])
peak_movies = int(peak_row['Number of Movies'])

avg_movies_per_year = int(movies_per_year['Number of Movies'].mean())

# -----------------------------
# Create Interactive Chart
# -----------------------------

fig = px.line(
    movies_per_year,
    x='Year',
    y='Number of Movies',
    title="<b>IMDb</b> | Number of Movies/Series Released Per Year",
    markers=True
)

# IMDb-inspired styling
fig.update_layout(
    template="plotly_dark",
    title_font=dict(size=22,color="#F5C518"),
    xaxis_title="Year",
    yaxis_title="Number of Movies/Series",
    hovermode="x unified",

    # create extra space on the right
    margin=dict(r=250)
)

# IMDb yellow line
fig.update_traces(
    line=dict(color="#F5C518", width=3),
    marker=dict(color="#F5C518", size=6)
)

# -----------------------------
# Insight Annotation Box
# -----------------------------

insight_text = (
    f"<b>Key Insights</b><br>"
    f"• {total_movies:,} movies released<br>"
    f"• Peak production: {peak_movies} movies/series in {peak_year}<br>"
    f"• Avg production: {avg_movies_per_year} movies/year"
)

fig.add_annotation(
    xref="paper",
    yref="paper",
    x=0.15,   # moved further right
    y=0.95,
    showarrow=False,
    align="left",
    bgcolor="rgba(0,0,0,0.7)",
    bordercolor="#F5C518",
    borderwidth=1,
    font=dict(size=12),
    text=insight_text
)

fig.show()

#### 3.3 Are there periods showing growth or decline in movie/series releases?

In [21]:
# @title
import pandas as pd
import plotly.express as px

# ---------------------------------------------------
# Data Preparation
# ---------------------------------------------------

filtered_df = df[
    (df['start_year'] > 1900) &
    (df['start_year'] <= 2020)
]

movies_per_year = (
    filtered_df['start_year']
    .value_counts()
    .sort_index()
    .reset_index()
)

movies_per_year.columns = ['Year', 'Number of Movies']

# ---------------------------------------------------
# Calculate Growth / Decline
# ---------------------------------------------------

movies_per_year['Yearly Change'] = (
    movies_per_year['Number of Movies']
    .diff()
)

# Highest growth year
growth_row = movies_per_year.loc[
    movies_per_year['Yearly Change'].idxmax()
]

growth_year = int(growth_row['Year'])
growth_value = int(growth_row['Yearly Change'])

# Highest decline year
decline_row = movies_per_year.loc[
    movies_per_year['Yearly Change'].idxmin()
]

decline_year = int(decline_row['Year'])
decline_value = int(decline_row['Yearly Change'])

# Peak production year
peak_row = movies_per_year.loc[
    movies_per_year['Number of Movies'].idxmax()
]

peak_year = int(peak_row['Year'])
peak_movies = int(peak_row['Number of Movies'])

# ---------------------------------------------------
# Create Interactive Visualization
# ---------------------------------------------------

fig = px.line(
    movies_per_year,
    x='Year',
    y='Number of Movies',

    title="<b>IMDb</b> | Growth & Decline in Movie/Series Releases",

    markers=True
)

# ---------------------------------------------------
# IMDb-inspired Styling
# ---------------------------------------------------

fig.update_layout(
    template="plotly_dark",

    title_font=dict(
        size=24,
        color="#F5C518"
    ),

    xaxis_title="Year",
    yaxis_title="Number of Movies/Series Released",

    hovermode="x unified",

    width=1250,
    height=700,

    margin=dict(
        l=80,
        r=500,
        t=80,
        b=100
    )

)

# IMDb yellow styling
fig.update_traces(
    line=dict(
        color="#F5C518",
        width=3
    ),

    marker=dict(
        color="#F5C518",
        size=6
    )
)

# ---------------------------------------------------
# Highlight Growth Year
# ---------------------------------------------------

fig.add_scatter(
    x=[growth_year],
    y=[growth_row['Number of Movies']],

    mode='markers+text',

    text=[f"Highest Growth ({growth_year})"],

    textposition="top center",

    marker=dict(
        size=14,
        color="lime"
    ),

    name="Highest Growth"
)

# ---------------------------------------------------
# Highlight Decline Year
# ---------------------------------------------------

fig.add_scatter(
    x=[decline_year],
    y=[decline_row['Number of Movies']],

    mode='markers+text',

    text=[f"Largest Decline ({decline_year})"],

    textposition="bottom center",

    marker=dict(
        size=14,
        color="red"
    ),

    name="Largest Decline"
)

# ---------------------------------------------------
# Insight Annotation Box
# ---------------------------------------------------

insight_text = (
    f"<b>Key Insights</b><br>"
    f"• Peak production: {peak_movies} releases in {peak_year}<br>"
    f"• Highest yearly growth: +{growth_value} releases ({growth_year})<br>"
    f"• Largest yearly decline: {decline_value} releases ({decline_year})<br>"
    f"• Long-term production trends are clearly visible"
)

fig.add_annotation(
    xref="paper",
    yref="paper",

    x=0.5,
    y=0.5,

    showarrow=False,

    align="left",

    bgcolor="rgba(0,0,0,0.7)",

    bordercolor="#F5C518",
    borderwidth=1,

    font=dict(size=12),

    text=insight_text
)

# ---------------------------------------------------
# Show Visualization
# ---------------------------------------------------

fig.show()

#### 3.4 How has the average number of votes per movie/series changed over the years?
#### 3.5 Which years show the highest audience engagement based on votes?

In [22]:
# @title
df['votes'] = pd.to_numeric(df['votes'], errors='coerce')

result = (
    df.groupby('end_year')['votes']
      .mean()
      .reset_index(name='average_votes')
)

print(result)
result['average_votes'].max()

     end_year  average_votes
0           0       0.013137
1        1877      48.250000
2        1878     119.000000
3        1881      60.000000
4        1885     347.000000
..        ...            ...
140      2026       0.000000
141      2027       0.000000
142      2028       0.000000
143      2029       0.000000
144      2100       0.027583

[145 rows x 2 columns]


1114.0705009276437

In [23]:
# @title
import pandas as pd
import plotly.express as px

# ---------------------------------------------------
# Data Preparation
# ---------------------------------------------------

# Convert votes column to numeric
df['votes'] = pd.to_numeric(df['votes'], errors='coerce')

# Keep only years between 1900 and 2020
filtered_df = df[
    (df['end_year'] >= 1900) &
    (df['end_year'] <= 2020)
]

# Calculate average votes per year
result = (
    filtered_df
    .groupby('end_year')['votes']
    .mean()
    .reset_index(name='average_votes')
)

# Rename columns for cleaner plotting
result.columns = ['Year', 'Average Votes']

# ---------------------------------------------------
# Calculate Insights
# ---------------------------------------------------

peak_row = result.loc[result['Average Votes'].idxmax()]

peak_year = int(peak_row['Year'])
peak_votes = int(peak_row['Average Votes'])

avg_votes_overall = int(result['Average Votes'].mean())

first_year = int(result['Year'].min())
last_year = int(result['Year'].max())

# ---------------------------------------------------
# Create Interactive Visualization
# ---------------------------------------------------

fig = px.line(
    result,
    x='Year',
    y='Average Votes',
    title="<b>IMDb</b> | Average Audience Votes Per Movie/Series Over Time",
    markers=True
)

# IMDb-inspired styling
fig.update_layout(
    template="plotly_dark",

    title_font=dict(
        size=24,
        color="#F5C518"
    ),

    xaxis_title="Year",
    yaxis_title="Average Votes",

    hovermode="x unified",

    margin=dict(r=260),

    xaxis=dict(
        range=[1900, 2020]
    )
)

# IMDb yellow styling
fig.update_traces(
    line=dict(
        color="#F5C518",
        width=3
    ),

    marker=dict(
        color="#F5C518",
        size=6
    )
)

# ---------------------------------------------------
# Add Insight Annotation Box
# ---------------------------------------------------

insight_text = (
    f"<b>Key Insights</b><br>"
    f"• Data range: {first_year} - {last_year}<br>"
    f"• Highest engagement: {peak_year}<br>"
    f"• Peak avg votes: {peak_votes:,}<br>"
    f"• Overall avg votes/year: {avg_votes_overall:,}"
)

fig.add_annotation(
    xref="paper",
    yref="paper",

    x=1.02,
    y=0.95,

    showarrow=False,

    align="left",

    bgcolor="rgba(0,0,0,0.7)",

    bordercolor="#F5C518",
    borderwidth=1,

    font=dict(
        size=12
    ),

    text=insight_text
)

# ---------------------------------------------------
# Highlight Peak Engagement Year
# ---------------------------------------------------

fig.add_scatter(
    x=[peak_year],
    y=[peak_votes],

    mode='markers+text',

    text=[f"Peak Engagement ({peak_year})"],

    textposition="top center",

    marker=dict(
        size=14,
        color="red"
    ),

    name="Peak Engagement"
)

# ---------------------------------------------------
# Show Chart
# ---------------------------------------------------

fig.show()

#### 3.6 How have average movie/series ratings changed over time?

In [24]:
# @title
result = (
    df[df['rating'] <= 10]      # keep ratings 10 or below
    .groupby('end_year')['rating']    # group by year
    .mean()                       # calculate average rating
    .reset_index(name='average_rating')
)


#df_



In [25]:
# @title
import pandas as pd
import plotly.express as px

# ---------------------------------------------------
# Data Preparation
# ---------------------------------------------------

# Convert rating column to numeric
df['rating'] = pd.to_numeric(df['rating'], errors='coerce')

# Keep valid ratings and years between 1900–2020
filtered_df = df[
    (df['rating'] <= 10) &
    (df['end_year'] >= 1900) &
    (df['end_year'] <= 2020)
]

# Calculate average rating per year
result = (
    filtered_df
    .groupby('end_year')['rating']
    .mean()
    .reset_index(name='average_rating')
)

# Rename columns for cleaner plotting
result.columns = ['Year', 'Average Rating']

# ---------------------------------------------------
# Calculate Insights
# ---------------------------------------------------

highest_row = result.loc[result['Average Rating'].idxmax()]
lowest_row = result.loc[result['Average Rating'].idxmin()]

highest_year = int(highest_row['Year'])
highest_rating = round(highest_row['Average Rating'], 2)

lowest_year = int(lowest_row['Year'])
lowest_rating = round(lowest_row['Average Rating'], 2)

overall_avg_rating = round(result['Average Rating'].mean(), 2)

first_year = int(result['Year'].min())
last_year = int(result['Year'].max())

# ---------------------------------------------------
# Create Interactive Visualization
# ---------------------------------------------------

fig = px.line(
    result,
    x='Year',
    y='Average Rating',
    title="<b>IMDb</b> | Average Movie/Series Ratings Over Time",
    markers=True
)

# IMDb-inspired styling
fig.update_layout(
    template="plotly_dark",

    title_font=dict(
        size=24,
        color="#F5C518"
    ),

    xaxis_title="Year",
    yaxis_title="Average IMDb Rating",

    hovermode="x unified",

    margin=dict(r=260),

    xaxis=dict(
        range=[1900, 2020]
    )
)

# IMDb yellow styling
fig.update_traces(
    line=dict(
        color="#F5C518",
        width=3
    ),

    marker=dict(
        color="#F5C518",
        size=6
    )
)

# ---------------------------------------------------
# Add Insight Annotation Box
# ---------------------------------------------------

insight_text = (
    f"<b>Key Insights</b><br>"
    f"• Data range: {first_year} - {last_year}<br>"
    f"• Highest avg rating: {highest_rating} ({highest_year})<br>"
    f"• Lowest avg rating: {lowest_rating} ({lowest_year})<br>"
    f"• Overall avg IMDb rating: {overall_avg_rating}"
)

fig.add_annotation(
    xref="paper",
    yref="paper",

    x=1.18,   # move further right
    y=0.5,    # vertically centered

    showarrow=False,

    align="left",

    bgcolor="rgba(0,0,0,0.7)",

    bordercolor="#F5C518",
    borderwidth=1,

    font=dict(
        size=12
    ),

    text=insight_text
)
# ---------------------------------------------------
# Highlight Highest Rated Year
# ---------------------------------------------------

fig.add_scatter(
    x=[highest_year],
    y=[highest_rating],

    mode='markers+text',

    text=[f"Highest Rated Year ({highest_year})"],

    textposition="top center",

    marker=dict(
        size=14,
        color="red"
    ),

    name="Highest Rated Year"
)

# ---------------------------------------------------
# Show Chart
# ---------------------------------------------------

fig.show()

#### 3.7 Are newer movies/series receiving higher or lower ratings compared to older movies?

In [26]:
# @title
import pandas as pd
import plotly.graph_objects as go

# ---------------------------------------------------
# Data Preparation
# ---------------------------------------------------

# Convert ratings to numeric
df['rating'] = pd.to_numeric(df['rating'], errors='coerce')

# Keep valid ratings only
filtered_df = df[
    (df['rating'] <= 10) &
    (df['start_year'] >= 1900) &
    (df['start_year'] <= 2020)
]

# ---------------------------------------------------
# Separate Old vs New Eras
# ---------------------------------------------------

old_movies = filtered_df[
    (filtered_df['start_year'] >= 1900) &
    (filtered_df['start_year'] <= 1940)
]

new_movies = filtered_df[
    (filtered_df['start_year'] >= 1980) &
    (filtered_df['start_year'] <= 2020)
]

# Average ratings per year
old_trend = (
    old_movies
    .groupby('start_year')['rating']
    .mean()
    .reset_index(name='avg_rating')
)

new_trend = (
    new_movies
    .groupby('start_year')['rating']
    .mean()
    .reset_index(name='avg_rating')
)

# ---------------------------------------------------
# Calculate Insights
# ---------------------------------------------------

old_avg = round(old_trend['avg_rating'].mean(), 2)
new_avg = round(new_trend['avg_rating'].mean(), 2)

comparison = (
    "higher"
    if new_avg > old_avg
    else "lower"
)

# ---------------------------------------------------
# Create Visualization
# ---------------------------------------------------

fig = go.Figure()

# Old Movies Line
fig.add_trace(
    go.Scatter(
        x=old_trend['start_year'],
        y=old_trend['avg_rating'],
        mode='lines+markers',
        name='1900-1940 (Classic Era)',
        line=dict(color='#F5C518', width=3),
        marker=dict(size=6)
    )
)

# New Movies Line
fig.add_trace(
    go.Scatter(
        x=new_trend['start_year'],
        y=new_trend['avg_rating'],
        mode='lines+markers',
        name='1980-2020 (Modern Era)',
        line=dict(color='cyan', width=3),
        marker=dict(size=6)
    )
)

# ---------------------------------------------------
# Layout Styling
# ---------------------------------------------------

fig.update_layout(
    template='plotly_dark',

    title="<b>IMDb</b> | Classic vs Modern Movie Ratings",

    title_font=dict(
        size=24,
        color="#F5C518"
    ),

    xaxis_title="Release Year",
    yaxis_title="Average IMDb Rating",

    hovermode="x unified",

    margin=dict(r=350),

    legend=dict(
        x=0.02,
        y=0.98
    )
)

# ---------------------------------------------------
# Insight Annotation
# ---------------------------------------------------

insight_text = (
    f"<b>Key Insights</b><br>"
    f"• Classic Era Avg Rating: {old_avg}<br>"
    f"• Modern Era Avg Rating: {new_avg}<br>"
    f"• Modern movies generally receive <b>{comparison}</b> ratings<br>"
    f"• Comparison focuses on distinct historical periods"
)

fig.add_annotation(
    xref="paper",
    yref="paper",

    x=1.18,
    y=0.5,

    showarrow=False,

    align="left",

    bgcolor="rgba(0,0,0,0.7)",

    bordercolor="#F5C518",
    borderwidth=1,

    font=dict(size=12),

    text=insight_text
)

# ---------------------------------------------------
# Show Chart
# ---------------------------------------------------

fig.show()

#### 3.8 Which genres were most popular in different years?
#### 3.9 Are there noticeable shifts in genre popularity over time?

In [27]:
# @title
result = (
    df.groupby('end_year')['genre']
      .agg(lambda x: x.mode().iloc[0])
      .reset_index(name='most_popular_genre')
)

print(result)

     end_year          most_popular_genre
0           0                    Thriller
1        1877            Animation, Short
2        1878            Animation, Short
3        1881            Animation, Short
4        1885            Animation, Short
..        ...                         ...
140      2026                   Animation
141      2027  Action, Adventure, Fantasy
142      2028                      Horror
143      2029                      Action
144      2100                    Thriller

[145 rows x 2 columns]


In [28]:
# @title
import pandas as pd
import plotly.express as px

# ---------------------------------------------------
# Data Preparation
# ---------------------------------------------------

# Keep only valid years
filtered_df = df[
    (df['end_year'] >= 1900) &
    (df['end_year'] <= 2020)
].copy()

# Remove missing genres
filtered_df = filtered_df.dropna(subset=['genre'])

# ---------------------------------------------------
# Extract Primary Genre
# (Some IMDb datasets contain multiple genres)
# ---------------------------------------------------

filtered_df['primary_genre'] = (
    filtered_df['genre']
    .astype(str)
    .str.split(',')
    .str[0]
)

# ---------------------------------------------------
# Find Most Popular Genre Per Year
# ---------------------------------------------------

genre_trend = (
    filtered_df
    .groupby('end_year')['primary_genre']
    .agg(lambda x: x.mode().iloc[0])
    .reset_index(name='most_popular_genre')
)

genre_trend.columns = ['Year', 'Most Popular Genre']

# ---------------------------------------------------
# Count Genre Dominance
# ---------------------------------------------------

genre_counts = (
    genre_trend['Most Popular Genre']
    .value_counts()
    .reset_index()
)

genre_counts.columns = ['Genre', 'Years Dominated']

top_genre = genre_counts.iloc[0]['Genre']
top_count = genre_counts.iloc[0]['Years Dominated']

# ---------------------------------------------------
# Create Visualization
# ---------------------------------------------------

fig = px.scatter(
    genre_trend,
    x='Year',
    y='Most Popular Genre',
    color='Most Popular Genre',

    title="<b>IMDb</b> | Genre Popularity Shifts Over Time"
)

# ---------------------------------------------------
# IMDb-inspired Styling
# ---------------------------------------------------

fig.update_layout(
    template='plotly_dark',

    title_font=dict(
        size=24,
        color="#F5C518"
    ),

    xaxis_title="Year",
    yaxis_title="Most Popular Genre",

    hovermode="closest",

    margin=dict(r=350),

    legend_title="Genres"
)

# Larger markers
fig.update_traces(
    marker=dict(size=10)
)

# ---------------------------------------------------
# Add Insight Annotation
# ---------------------------------------------------

insight_text = (
    f"<b>Key Insights</b><br>"
    f"• Most dominant genre overall: <b>{top_genre}</b><br>"
    f"• Dominated for {top_count} years<br>"
    f"• Clear genre shifts visible across decades<br>"
    f"• Some genres rise and decline over time"
)

fig.add_annotation(
    xref="paper",
    yref="paper",

    x=1.8,   # further right
    y=-0.12,   # bottom

    showarrow=False,

    align="left",

    bgcolor="rgba(0,0,0,0.7)",

    bordercolor="#F5C518",
    borderwidth=1,

    font=dict(size=12),

    text=insight_text
)

# ---------------------------------------------------
# Show Visualization
# ---------------------------------------------------

fig.show()

#### 3.10)Has the average movie duration changed across years?

In [29]:
# @title
df['duration'] = pd.to_numeric(df['duration'], errors='coerce')

result = (
    df.groupby('end_year')['duration']
      .mean()
      .reset_index(name='average_duration')
)
result

,end_year,average_duration
0,0,4.562250
1,1877,1.000000
2,1878,1.000000
3,1881,1.000000
4,1885,1.000000
...,...,...
140,2026,9.047619
141,2027,19.409091
142,2028,0.000000
143,2029,0.000000


In [30]:
# @title
import pandas as pd
import plotly.express as px

# ---------------------------------------------------
# Data Preparation
# ---------------------------------------------------

# Convert duration column to numeric
df['duration'] = pd.to_numeric(df['duration'], errors='coerce')

# Filter years between 1900–2020
filtered_df = df[
    (df['end_year'] >= 1900) &
    (df['end_year'] <= 2020)
]

# Average duration per year
duration_trend = (
    filtered_df
    .groupby('end_year')['duration']
    .mean()
    .reset_index(name='average_duration')
)

duration_trend.columns = ['Year', 'Average Duration']

# ---------------------------------------------------
# Key Insights
# ---------------------------------------------------

highest_row = duration_trend.loc[
    duration_trend['Average Duration'].idxmax()
]

lowest_row = duration_trend.loc[
    duration_trend['Average Duration'].idxmin()
]

highest_year = int(highest_row['Year'])
highest_duration = round(highest_row['Average Duration'], 1)

lowest_year = int(lowest_row['Year'])
lowest_duration = round(lowest_row['Average Duration'], 1)

overall_avg = round(
    duration_trend['Average Duration'].mean(),
    1
)

# ---------------------------------------------------
# Create Visualization
# ---------------------------------------------------

fig = px.line(
    duration_trend,
    x='Year',
    y='Average Duration',

    title="<b>IMDb</b> | Average Movie/Series Duration Across Years",

    markers=True
)

# ---------------------------------------------------
# IMDb-inspired Styling
# ---------------------------------------------------

fig.update_layout(
    template="plotly_dark",

    title_font=dict(
        size=24,
        color="#F5C518"
    ),

    xaxis_title="Year",
    yaxis_title="Average Duration (Minutes)",

    hovermode="x unified",

    margin=dict(
        r=420,
        b=80
    ),

    xaxis=dict(
        range=[1900, 2020]
    )
)

# IMDb yellow styling
fig.update_traces(
    line=dict(
        color="#F5C518",
        width=3
    ),

    marker=dict(
        color="#F5C518",
        size=6
    )
)

# ---------------------------------------------------
# Insight Annotation
# ---------------------------------------------------

insight_text = (
    f"<b>Key Insights</b><br>"
    f"• Longest avg duration: {highest_duration} mins ({highest_year})<br>"
    f"• Shortest avg duration: {lowest_duration} mins ({lowest_year})<br>"
    f"• Overall avg duration: {overall_avg} mins<br>"
    f"• Runtime patterns vary noticeably across decades"
)

fig.add_annotation(
    xref="paper",
    yref="paper",

    x=2,
    y=0.5,

    showarrow=False,

    align="left",

    bgcolor="rgba(0,0,0,0.7)",

    bordercolor="#F5C518",
    borderwidth=1,

    font=dict(size=12),

    text=insight_text
)

# ---------------------------------------------------
# Show Visualization
# ---------------------------------------------------

fig.show()

## Final Findings & Business Impact

### Key Findings

- **52,219 movies/series** were analysed across **1900–2020+**.
- Content production peaked at **114,225 releases in 2017**, demonstrating significant industry expansion.
- The overall average IMDb rating was **6.42**, with the highest average rating of **7.25 in 1963**.
- Modern content outperformed classic content by **1.09 rating points** (**6.88 vs 5.79**).
- Peak audience engagement reached **1,114 average votes in 1942**.
- **Animation was the most dominant genre**, leading for approximately **52 years**.

### Business Impact

- **Scale alone does not guarantee success:** Production reached **114,225 releases at its peak**, but higher content volume does not automatically translate into proportional audience engagement. Companies should optimize for **quality and discoverability**, not volume alone.

- **Quality can create competitive advantage:** Modern content achieved an average rating **18.8% higher than classic-era content** (**6.88 vs 5.79**), highlighting the potential value of investing in stronger content quality and production standards.

- **Audience attention is highly concentrated:** Peak engagement reached **1,114 average votes**, showing that successful content can attract significantly greater audience attention than the average title. Effective marketing and recommendation systems can help maximize this opportunity.

- **Genre strategy must remain dynamic:** With **Animation dominating for approximately 52 years**, the analysis demonstrates that long-term genre trends can create major opportunities. Studios should continuously measure genre performance before allocating production budgets.

- **Data-driven targeting becomes increasingly important:** Across **52,219 analysed titles**, increasing content availability creates stronger competition for audience attention. Personalized recommendations, targeted marketing, and audience segmentation can improve **content discovery, engagement, and return on investment (ROI)**.

> **Business Takeaway:** As the industry grows beyond **52,000+ analysed titles**, the strongest competitive advantage is not producing the highest volume of content—it is using data to identify high-potential content, invest in quality, and deliver the right content to the right audience.
>
